### ISSUE TITLE
Have a look at dbaarda/rollsum-tests/ and librsync.

### ISSUE THREAD

**Issue by dbaarda** [aidanhs/rollsum-tests#2]
G'day,

I'm not sure of the best way to communicate with you about this, using this issue as a convenient way to send you a message. FTR I stumbled on this via the IPFS discussions on chunking.

I too did some analysis of various rollsums for librsync, the results of which are here;
 
https://github.com/dbaarda/rollsum-tests/blob/master/RESULTS.rst

I was looking at it from a librsync perspective, which uses it as a quick hash of a whole block, not a small window for chunking. This means I cared about hash distribution and collisions more that a chunking algorithm probably would.

I found RabinKarp (Polynomial rolling hash, not RabinFingerprint) to be the winner with the best distribution and low hash collisions, while also being very simple. It should also fare pretty well for chunking for similar reasons.

I didn't consider RabinFingerprint because I considered GF(2) maths to be unnecessarily complicated. I've since stumbled over a few articles/comments that suggest that crypto research's GF(2) obsession is because they come from a school focused on fast hardware implementations, and that it doesn't actually contribute to better security, just makes it slower in software. Hence fast (in software) secure hashes like blake2 use normal arithmetic instead. So in this respect RabinKarp (Polyhash) is the fast-software equivalent of Rabinfingerprint.

Librsync has fast C implementations of both rsync's rollsum (what you call bap) and RabinKarp (Polynomial rolling hash) here;

https://github.com/librsync/librsync/blob/master/src/rollsum.h
https://github.com/librsync/librsync/blob/master/src/rabinkarp.h

**Comment by dbaarda** [comment]
I've also just finished doing some more testing and analysis of chunking algorithms inspired by the FastCDC paper here;

https://github.com/dbaarda/rollsum-chunking/blob/master/RESULTS.rst

The results of this was a bit surprising: FastCDC is not better than simple exponential chunkers when using the same average and minimum chunk size. The best minimum to use for simple exponential chunkers is much larger than what is normally recommended.

**Comment by aidanhs** [comment]
Ha, it looks like I'd already discovered your repo in Feb of 2020 (https://github.com/aidanhs/rollsum-tests/issues/1) a month before you created this one.

First off, I just want to say that your analysis is awesome. I particularly love that you've assessed some of the mathematical properties of the algorithms and how that influences viable use-cases.

I think it'd be cool to have a prominent link in the readme of my repo to refer interested readers to your repo (or this issue if you think it more appropriate) if they want different kinds of analysis - what do you think? Would you be up for creating a PR that would do your repo justice? 😄

**Comment by dbaarda** [comment]
Sure, I'll try to find the time to send you a PR (hmm, what happens on github if I fork someone else's repo that has the same name as one of mine). I'll also try to add some links to yours on mine.

I also occasionally update these with more info. I think I added gear to my rollsum-results since I posted here, and I have some changes coming to the rollsum-chunking that includes "Regression Chunking" that comes from a Microsoft paper on chunking.
